### PR TITLE
refactor(data-access): thegraph-data-access package

### PR DESCRIPTION
**Pull request by benjlevesque**
# Description
Introduce a new package to easily re-use thegraph data access. 
Move non-graph logic to their relevant places

# Usage

## Full 
```ts
import { RequestNetworkBase } from '@requestnetwork/request-client.js';
import { TheGraphDataAccess } from '@requestnetwork/thegraph-data-access';
import { Wallet } from 'ethers';
import { IpfsStorage, EthereumStorageEthers } from '@requestnetwork/ethereum-storage';

const signer = Wallet.createRandom();
const storage = new EthereumStorageEthers({
  ipfsStorage: new IpfsStorage({}),
  signer,
  network: 'private',
});
const dataAccess = new TheGraphDataAccess({
  graphql: { url: 'http://localhost:8000/subgraphs/name/RequestNetwork/request-storage' },
  network: 'private',
  storage,
});
const rn = new RequestNetworkBase({ dataAccess });

const { requestId } = await rn.createRequest(...)

const request = await rn.fromRequestId(requestId);

await request.accept(...)
```

## Read only
```ts
import { RequestNetworkBase } from '@requestnetwork/request-client.js';
import { TheGraphDataAccess } from '@requestnetwork/thegraph-data-access';
import { Wallet } from 'ethers';

const dataAccess = new TheGraphDataAccess({
  graphql: { url: 'http://localhost:8000/subgraphs/name/RequestNetwork/request-storage' },
  network: 'private',
});
const rn = new RequestNetworkBase({ dataAccess });

// Cannot create request
// await rn.createRequest()

const request = await rn.fromRequestId('...');
console.log(request.requestId);

// Cannot modify request
// await request.accept(...)
```

### CODE REVIEW

File: packages/request-node/src/thegraph-node.ts [hunk 1]
```diff
@@ -25,12 +28,17 @@ export class TheGraphRequestNode extends RequestNodeBase {
...
       graphql: { url },
-      ipfsStorage,
+      storage,
       network,
-      signer,
-      logger: logger,
+      logger,
     });
 
     super(dataAccess, ipfsStorage, store, logger);
```
Comment (yomarion) [inline packages/request-node/src/thegraph-node.ts:44]: Have you kept `ipfsStorage` on purpose? It looks weird so maybe add a comment to explain

File: packages/thegraph-data-access/.nycrc [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (yomarion) [inline packages/thegraph-data-access/.nycrc:1]: Is this file versioned on purpose?

File: packages/thegraph-data-access/src/data-access.ts [hunk 3]
```diff
@@ -8,20 +8,25 @@ import { Block } from '@requestnetwork/data-access';
...
-  ipfsStorage: StorageTypes.IIpfsStorage;
-  graphql: { url: string } & RequestInit;
-  signer: Signer;
+type TheGraphDataAccessBaseOptions = {
   network: string;
   logger?: LogTypes.ILogger;
   pendingStore?: PendingStore;
 };
 
+export type TheGraphDataAccessOptions = TheGraphDataAccessBaseOptions & {
```
Comment (yomarion) [inline packages/thegraph-data-access/src/data-access.ts:21]:
```suggestion
export type TheGraphDataReadOptions = TheGraphDataAccessBaseOptions & {
```
Might make the difference with the other type more explicit?

Comment (yomarion) [reply packages/thegraph-data-access/src/data-access.ts:21]: It's also consistent with the  `TheGraphDataRead` class name

File: packages/types/src/storage-types.ts [hunk 4]
```diff
@@ -2,14 +2,21 @@ import { EventEmitter } from 'events';
...
+export interface IStorageRead {
+  initialize: () => Promise<void>;
   read: (dataId: string) => Promise<IEntry>;
   readMany: (dataIds: string[]) => Promise<IEntry[]>;
   getData: (options?: ITimestampBoundaries) => Promise<IEntriesWithLastTimestamp>;
   getIgnoredData: () => Promise<IEntry[]>;
+}
+
+/** Interface of the storage */
+export interface IStorage extends IStorageRead, IStorageWrite {
```
Comment (yomarion) [inline packages/types/src/storage-types.ts:19]: 💅

**Review by yomarion — APPROVED** [review]

File: packages/thegraph-data-access/src/data-access.ts (hunk 3 above)
Comment (benjlevesque) [reply packages/thegraph-data-access/src/data-access.ts:21]: i removed datawrite and moved the `storage` to `TheGraphDataAccessOptions`, which now makes sense like this.

File: packages/thegraph-data-access/.nycrc (hunk 2 above)
Comment (benjlevesque) [reply packages/thegraph-data-access/.nycrc:1]: yes, we have this on every package... nyc is the tool for test coverage

**Review by alexandre-abrioux — COMMENTED** [review]
Nice 😀

File: packages/request-node/src/thegraph-node.ts (hunk 1 above)
Comment (alexandre-abrioux) [reply packages/request-node/src/thegraph-node.ts:44]: @benjlevesque I understand what @yomarion means (I think). I agree that it's weird to have an `ipfsStorage` object in the `RequestNodeBase` constructor, knowing that we pass down a `dataAccess` object that should already be a layer on top of IPFS. The issue is that it doesn't look possible to abstract—in the data-access layer—what `ipfsStorage` is used for in the request-node: the `/ipfsAdd` endpoint. This endpoint only persists documents on IPFS without persisting anything on-chain. I don't think it makes sense to add a generic method in data-access that only persist documents off-chain (like `persistTransactionOffChain` for instance). The data-access interface shouldn't contain methods that are tied to the underlying technology used for persisting IMO. Honestly, I think this endpoint is a security hole anyway, allowing users to persist documents on IPFS without persisting anything on-chain means we disregard spamming protection and the IPFS node could quickly become filled with fake documents. I'd propose we completely remove that endpoint, I don't think it is used anywhere, right?

Comment (benjlevesque) [reply packages/request-node/src/thegraph-node.ts:44]: Yeah sorry we solved this in DM, my bad. 

this endpoint should be challenged for sure. It is related to the option I removed from the client. I believe it _should_ be possible to use 100% the client as a front-end library, but it doesn't need to be first-class like it is today

Comment (alexandre-abrioux) [reply packages/request-node/src/thegraph-node.ts:44]: Ha! Get it, thanks!

**Pull request merged**